store/storethehash/storethehash.go

```diff
@@ -78,48 +85,45 @@ func (s *sthStorage) get(k []byte) ([]indexer.Value, bool, error) {
 
 }
 
-func (s *sthStorage) ForEach(iterFunc indexer.IterFunc) error {
+func (s *sthStorage) Iter() (indexer.Iterator, error) {
 	s.Flush()
 	iter, err := s.primary.Iter()
 	if err != nil {
-		return err
+		return nil, err
 	}
+	return &sthIterator{
+		iter:     iter,
+		storage:  s,
+		uniqKeys: map[string]struct{}{},
+	}, nil
+}
 
-	// Create a list of unique keys, and then call get for each unique key.
-	//
-	// This is necessary because the primary iterator returns all versions of
-	// each index.  If a key has been updated with new values, the iterator
-	// returns the key with the original values, then again with the values for
-	// the next update, and so on for as many updates a there were for a key.
-	uniqKeys := map[string]struct{}{}
+func (it *sthIterator) Next() (multihash.Multihash, []indexer.Value, error) {
 	for {
-		key, _, err := iter.Next()
+		key, _, err := it.iter.Next()
 		if err != nil {
 			if err == io.EOF {
-				break
+				it.uniqKeys = nil
 			}
-			return err
+			return nil, nil, err
 		}
+
 		k := string(key)
-		if _, found := uniqKeys[k]; found {
+		_, found := it.uniqKeys[k]
```

Comment:
We should maybe add an issue to go-storethehash to suggest implementing in the iterator a way to look to the freelist and go through outdated slots of memory as discussed. Otherwise, LGTM

Reply:
Added here: https://github.com/ipld/go-storethehash/issues/14